CDP4JsonSerializer/AutoGenDtoSerializer/PersonPermissionSerializer.cs

```diff
@@ -30,87 +29,429 @@
 {
     using System;
-    using System.Collections;
     using System.Collections.Generic;
     using System.Linq;
+    using System.Text.Json;
 
-    using CDP4Common.DTO;
+    using CDP4Common;
+    using CDP4Common.CommonData;
+    using CDP4Common.EngineeringModelData;
+    using CDP4Common.ReportingData;
+    using CDP4Common.SiteDirectoryData;
     using CDP4Common.Types;
 
-    using Newtonsoft.Json.Linq;
+    using NLog;
+
+    using Thing = CDP4Common.DTO.Thing;
+    using PersonPermission = CDP4Common.DTO.PersonPermission;
 
     /// <summary>
     /// The purpose of the <see cref="PersonPermissionSerializer"/> class is to provide a <see cref="PersonPermission"/> specific serializer
     /// </summary>
     public class PersonPermissionSerializer : BaseThingSerializer, IThingSerializer
     {
         /// <summary>
-        /// The map containing the serialization methods
+        /// Serializes a <see cref="Thing" /> into an <see cref="Utf8JsonWriter" />
         /// </summary>
-        private readonly Dictionary<string, Func<object, JToken>> propertySerializerMap = new Dictionary<string, Func<object, JToken>>
+        /// <param name="thing">The <see cref="Thing" /> that have to be serialized</param>
+        /// <param name="writer">The <see cref="Utf8JsonWriter" /></param>
+        /// <param name="requestedDataModelVersion">The <see cref="Version" /> that has been requested for the serialization</param>
+        /// <exception cref="ArgumentException">If the provided <paramref name="thing" /> is not an <see cref="PersonPermission" /></exception>
+        /// <exception cref="NotSupportedException">If the provided <paramref name="requestedDataModelVersion" /> is not supported</exception>
+        public void Serialize(Thing thing, Utf8JsonWriter writer, Version requestedDataModelVersion)
         {
-            { "accessRight", accessRight => new JValue(accessRight.ToString()) },
-            { "actor", actor => new JValue(actor) },
-            { "classKind", classKind => new JValue(classKind.ToString()) },
-            { "excludedDomain", excludedDomain => new JArray(excludedDomain) },
-            { "excludedPerson", excludedPerson => new JArray(excludedPerson) },
-            { "iid", iid => new JValue(iid) },
-            { "isDeprecated", isDeprecated => new JValue(isDeprecated) },
-            { "modifiedOn", modifiedOn => new JValue(((DateTime)modifiedOn).ToString("yyyy-MM-ddTHH:mm:ss.fffZ")) },
-            { "objectClass", objectClass => new JValue(objectClass.ToString()) },
-            { "revisionNumber", revisionNumber => new JValue(revisionNumber) },
-            { "thingPreference", thingPreference => new JValue(thingPreference) },
-        };
+            if (thing is not PersonPermission personPermission)
+            {
+                throw new ArgumentException("The thing shall be a PersonPermission", nameof(thing));
+            }
+
+            if (requestedDataModelVersion < Version.Parse("1.0.0"))

```

Comment:
can this ever happen, there is no version than 1.0.0. I think we should throw an exception since something is wrong

Reply:
This was probably introduced for objects that have a minimum version > 1.0.0. Basically all CDP4 Extensions. Code should throw an error if the requestedDataModelVersion is < 1.0.0

Reply:
resolved